pep-9999.rst

```diff
@@ -0,0 +1,850 @@
+PEP: 9999
+Title: Deep Immutability in Python
+Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>
+Sponsor: TBD
+Discussions-To: TBD
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 27-Feb-2025
+Python-Version: 3.15
+Post-History: 
+Resolution:
+
+
+Abstract
+========
+
+This PEP proposes adding a mechanism for deep immutability to
+Python. The mechanism requires some changes to the core language,
+but user-facing functions are delivered in a module called
+``freeze``. This module provides two functions:
+
+1. ``freeze(obj)`` -- turns ``obj`` deeply immutable
+2. ``is_immutable(obj)`` -- returns ``True`` if ``obj`` is immutable
+
+Making an object *deeply* immutable recursively renders the object
+*and all objects it references* immutable. (Just
+making the first object immutable is called *shallow*
+immutability.)
+
+Deep immutability provides strong guarantees against
+unintended modifications, thereby improving correctness, security, and
+parallel execution safety.
+
+Immutable objects are not immortal -- their reference counts
+stay mutable.
+In this PEP, we rely on the GIL to ensure the correctness of
+reference counts of immutable objects, but we have several planned
+memory management extensions, including support for atomic
+reference counting on immutable objects. These are outlined at the
+end of this document.
+
+Immutability in action:
+
+.. code-block:: python
+
+  from freeze import freeze, is_immutable
+
+  class Foo:
+      pass
+
+  f = Foo()
+  g = Foo()
+  h = Foo()
+
+  f.f = g
+  g.f = h
+  h.f = g # cycles are OK!
+  del g # Remove local ref to g, so g's RC = 1
+  del h # Remove local reg to h, so h's RC = 1
+
+  g.x = "African Swallow" # OK
+  freeze(f) # Makes, f, g and h immutable
+  g.x = "European Swallow" # Throws an exception "g is immutable"
+  is_immutable(h) # returns True
+  h = None # Cycle detector will eventually find and collect the cycle
+
+
+Motivation
+==========
+
+
+Ensuring Data Integrity
+-----------------------
+
+Python programs frequently manipulate large, interconnected data
+structures such as dictionaries, lists, and user-defined objects.
+Unintentional mutations can introduce subtle and
+difficult-to-debug errors. By allowing developers to explicitly
+freeze objects and their transitive dependencies, Python can
+provide stronger correctness guarantees for data processing
+pipelines, functional programming paradigms, and API boundaries
+where immutability is beneficial.
```

Comment:
Should we maybe mention caching here? So the fact that objects returned from a cache are usually mutable which can lead to interesting results if they changed?